src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -0,0 +1,384 @@
+                        safeWaitHandles[i] = null;
+                    }
+                }
+            }
+        }
+
+        private static int WaitMultiple(ReadOnlySpan<WaitHandle> waitHandles, bool waitAll, int millisecondsTimeout)
+        {
+            if (waitHandles == null)
+            {
+                throw new ArgumentNullException(nameof(waitHandles), SR.ArgumentNull_Waithandles);
+            }
+            if (waitHandles.Length == 0)
+            {
+                //
+                // Some history: in CLR 1.0 and 1.1, we threw ArgumentException in this case, which was correct.
+                // Somehow, in 2.0, this became ArgumentNullException.  This was not fixed until Silverlight 2,
+                // which went back to ArgumentException.
+                //
+                // Now we're in a bit of a bind.  Backward-compatibility requires us to keep throwing ArgumentException
+                // in CoreCLR, and ArgumentNullException in the desktop CLR.  This is ugly, but so is breaking
+                // user code.
+                //
+                throw new ArgumentException(SR.Argument_EmptyWaithandleArray, nameof(waitHandles));
+            }
+            if (waitHandles.Length > MaxWaitHandles)
+            {
+                throw new NotSupportedException(SR.NotSupported_MaxWaitHandles);
+            }
+            if (-1 > millisecondsTimeout)
+            {
+                throw new ArgumentOutOfRangeException(nameof(millisecondsTimeout), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
+            }
+
+            ObtainSafeWaitHandles(waitHandles, out SafeWaitHandle[] safeWaitHandles, out IntPtr[] unsafeWaitHandles);
+            try
+            {
+                int waitResult;
+
+                SynchronizationContext context = SynchronizationContext.Current;
+                if (context != null && context.IsWaitNotificationRequired())
+                {
+                    waitResult = context.Wait(unsafeWaitHandles, waitAll, millisecondsTimeout);
+                }
+                else
+                {
+                    waitResult = WaitMultipleIgnoringSyncContext(unsafeWaitHandles, waitAll, millisecondsTimeout);
+                }
+
+                if (waitResult >= WaitAbandoned && waitResult < WaitAbandoned + unsafeWaitHandles.Length)
+                {
+                    if (waitAll)
+                    {
+                        // In the case of WaitAll the OS will only provide the information that mutex was abandoned.
+                        // It won't tell us which one.  So we can't set the Index or provide access to the Mutex
+                        throw new AbandonedMutexException();
+                    }
+
+                    waitResult -= WaitAbandoned;
+                    throw new AbandonedMutexException(waitResult, waitHandles[waitResult]);
+                }
+
+                GC.KeepAlive(unsafeWaitHandles);
```

Comment:
Why is it needed here?

Reply:
I'm 99% sure it's not needed anymore, but I kept it from the original code. `WaitMultipleIgnoringSyncContext` keeps the reference to its parameters in GC-visible way as far as I can tell and the rest of code should not depend on it. (Same for the other `GC.KeepAlive`)

Reply:
I removed it, let's see if anything breaks.